### PR TITLE
Puzzles

### PR DESCRIPTION
**Pull request by domino14**

### CODE REVIEW

File: puzzles/puzzles.go [hunk 1]
```diff
@@ -124,6 +135,91 @@ func CELOnlyPuzzle(g *game.Game, moves []*move.Move) (bool, pb.PuzzleTag) {
...
+			return -1, fmt.Errorf("invalid puzzle bucket, tag %s appears more than once", tag.String())
+		}
+		tagValidationMap[tag] = true
+		res += intPow(3, int(tag)) * 1
+	}
+	return res, nil
+}
+
+// XXX: Should be moved to some common/utilities package
+func intPow(n, m int) int {
```
Comment (domino14) [inline puzzles/puzzles.go:212]: make it int64 to make it hold enough tags

File: puzzles/puzzles.go [hunk 2]
```diff
@@ -56,19 +66,135 @@ func CreatePuzzlesFromGame(conf *config.Config, g *game.Game) ([]*pb.PuzzleCreat
...
+	return moveIsBingo(m) && moveContainsBlank(m), pb.PuzzleTag_BLANK_BINGO
+}
+
+func NonBingoPuzzle(g *game.Game, moves []*move.Move) (bool, pb.PuzzleTag) {
+	return !moveIsBingo(moves[0]), pb.PuzzleTag_NON_BINGO
+}
+
+// XXX: Must be expanded to other languages
+func PowerTilePuzzle(g *game.Game, moves []*move.Move) (bool, pb.PuzzleTag) {
+	return moveContainsLetter(moves[0], "XJZQ"), pb.PuzzleTag_POWER_TILE
```
Comment (domino14) [inline puzzles/puzzles.go:113]: how about we look at the game's alphabet, and figure out from there if it has a letter that is worth `X` or more points (6, perhaps? 8?)

File: puzzles/puzzles.go [hunk 3]
```diff
@@ -124,6 +135,91 @@ func CELOnlyPuzzle(g *game.Game, moves []*move.Move) (bool, pb.PuzzleTag) {
...
+	if req.Buckets == nil {
+		return errors.New("buckets are nil in puzzle generation request")
+	}
+	bucketEncryptions := map[int]bool{}
+	for idx, bucket := range req.Buckets {
+		be, err := validatePuzzleBucket(bucket)
+		if err != nil {
+			return fmt.Errorf("error for bucket %d: %s", idx, err.Error())
+		}
+		if bucketEncryptions[be] {
```
Comment (domino14) [inline puzzles/puzzles.go:176]: should you modify `bucketEncryptions` in this loop as well?

Comment (jvc56) [reply puzzles/puzzles.go:187]: I don't understand. This is just a validation. What would we need to modify?

Comment (domino14) [reply puzzles/puzzles.go:187]: i have no idea what this was. i think you can ignore it 😅

Comment (domino14) [reply puzzles/puzzles.go:187]: i have no idea what this was. i think you can ignore it 😅

**Pull request merged**